lib/model/queue.go

```diff
@@ -7,65 +7,108 @@
 package model
 
 import (
+	"encoding/binary"
 	"math/rand"
-	"sort"
 	"time"
 
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/diskoverflow"
 	"github.com/syncthing/syncthing/lib/sync"
 )
 
 type jobQueue struct {
-	progress []string
-	queued   []jobQueueEntry
-	mut      sync.Mutex
-}
-
-type jobQueueEntry struct {
-	name     string
-	size     int64
-	modified time.Time
-}
-
-func newJobQueue() *jobQueue {
-	return &jobQueue{
-		mut: sync.NewMutex(),
+	progress       []string
+	queued         diskoverflow.Sorted
+	broughtToFront []string
+	handledAtFront map[string]struct{}
+	location       string
+	order          config.PullOrder
+	shuffleKeys    map[uint64]struct{}
+	mut            sync.Mutex
+}
+
+func newJobQueue(order config.PullOrder, loc string) *jobQueue {
+	q := &jobQueue{
+		handledAtFront: make(map[string]struct{}),
+		location:       loc,
+		order:          order,
+		mut:            sync.NewMutex(),
 	}
+	q.queued = diskoverflow.NewSorted(loc)
+	if order == config.OrderRandom {
+		q.shuffleKeys = make(map[uint64]struct{})
+	}
+	return q
 }
 
 func (q *jobQueue) Push(file string, size int64, modified time.Time) {
+	var key []byte
+	switch q.order {
+	case config.OrderRandom:
+		var n uint64
+		for {
+			n = rand.Uint64()
+			if _, ok := q.shuffleKeys[n]; !ok {
+				q.shuffleKeys[n] = struct{}{}
+				break
+			}
+		}
```

Comment:
So you spill to disk for values, but now blow up in memory to store the shuffle keys.

Reply:
Perhaps a bloom filter would work here?

Reply:
That's outdated, I can just put in random numbers: If equal ones occur, that's not an issue as `diskoverflow.Sorted` already has to take care that it doesn't overwrite stuff with the same key (it appends a sequence number). I don't know bloom filters, so if that's better/simpler, feel free to tell me to look it up (might take a while though).